Add prog Attach/Detach

florianl opened the pull request:
Signed-off-by: Lehner Florian <dev@der-flo.net>

This PR adds the syscall wrappers to attach/detach a program as mentioned in https://github.com/cilium/ebpf/issues/22.

lmb commented on prog.go:
```diff
@@ -413,6 +413,16 @@ func (p *Program) MarshalBinary() ([]byte, error) {
 }
 
+// Attach a Program to a container object fd
+func (p *Program) Attach(fd uint32, bfpProgType ProgramType, flags uint32) error {
```
`fd` should be `int`, and then we should check for `fd < 0`, to be consistent with other fds in the API. I think `flags` could be `type AttachFlags uint32`. We don't need to define all the flags now, but we leave this option open to us. User code can do `foo = ebpf.AttachFlags(0xbar)` in the meantime.

I'm a bit fuzzy what `bpfProgType` is for. Shouldn't it be `AttachType`?

florianl replied:
You are correct, it should be `AttachType` - sorry for the mistake!

lmb commented on prog.go:
```diff
@@ -413,6 +413,16 @@ func (p *Program) MarshalBinary() ([]byte, error) {
+}
+
+// Detach a Program from a container object fd
+func (p *Program) Detach(fd uint32, bfpProgType ProgramType, flags uint32) error {
```
Same comments as above.

florianl commented on syscalls.go:
```diff
@@ -205,6 +212,50 @@ func bpfProgLoad(attr *bpfProgLoadAttr) (*bpfFD, error) {
+
+	for {
+		_, err := bpfCall(_ProgAttach, unsafe.Pointer(&attr), unsafe.Sizeof(attr))
+		if err == unix.EAGAIN {
```
I also don't know of a case, where attach/detach calls can be interrupted. As I was not sure, I did it this way.
But I'll simplify it, as you suggest.

florianl commented on syscalls.go:
```diff
@@ -205,6 +212,50 @@ func bpfProgLoad(attr *bpfProgLoadAttr) (*bpfFD, error) {
 	}
 }
 
+func bpfProgAttach(m *bpfFD, targetFd, alterType, alterFlags uint32) error {
```
I was not sure how to call it. `bpfProgAlter(cmd, *bpfProgAlterAttr)` sounds good for me and I will update it.

florianl commented:
In the updated PR I've addressed the issues as suggested.

For the test, I'm still trying to find a solution. Most use cases of BPF_PROG_ATTACH I'm aware of, are in combination with cgroups. But I'm not sure if it is good to add such a dependency just for tests.

florianl commented:
The just added test case I heavily inspired by https://blog.cloudflare.com/sockmap-tcp-splicing-of-the-future/.

AkihiroSuda commented on syscalls.go:
```diff
@@ -173,6 +173,13 @@ type bpfProgTestRunAttr struct {
 
+type bpfProgAlterAttr struct {
+	targetFd   uint32
+	alterBpfFd uint32
```
s/alter/attacj/g ?
https://github.com/torvalds/linux/blob/v5.3/include/uapi/linux/bpf.h#L423-L427

florianl replied:
as the same struct is used for attaching and detaching, I've changes it to alter. But I will revert it, to keep it the way it is in the kernel.

florianl commented on prog.go:
```diff
@@ -414,13 +414,45 @@ func (p *Program) MarshalBinary() ([]byte, error) {
 // Attach a Program to a container object fd
-func (p *Program) Attach(fd uint32, bfpProgType ProgramType, flags uint32) error {
-	return bpfProgAttach(p.fd, fd, uint32(bfpProgType), flags)
+func (p *Program) Attach(fd int, attachType AttachType, attachFlags AttachFlags) error {
```
`type` is an identifier in the language. so, I wanted to avoid confusions here. any other suggestions for naming it?

AkihiroSuda replied:
`typ` seems the common convention to avoid conflict with `type`: https://golang.org/pkg/go/types/#IsInterface

lmb commented:
Thanks!
